src/model/src/SphericalJoint.cpp

```diff
@@ -0,0 +1,664 @@
+                             link1_X_link2_at_rest(_link1_X_link2)
+{
+    this->setPosCoordsOffset(0);
+    this->setDOFsOffset(0);
+
+    // Centers default to origin (no offset)
+    m_centerInFirstLink  = Position::Zero();
+    m_centerInSecondLink = Position::Zero();
+    m_centersExplicit    = false;
+
+    // Initialize with identity quaternion (w,x,y,z) = (1,0,0,0)
+    Vector4 identityQuat;
+    identityQuat(0) = 1.0; identityQuat(1) = 0.0; identityQuat(2) = 0.0; identityQuat(3) = 0.0;
+    this->resetAxisBuffers();
+    this->resetBuffers(identityQuat);
+}
+
+SphericalJoint::SphericalJoint(const SphericalJoint& other):
+                             link1(other.link1), link2(other.link2),
+                             link1_X_link2_at_rest(other.link1_X_link2_at_rest)
+{
+    this->setPosCoordsOffset(other.getPosCoordsOffset());
+    this->setDOFsOffset(other.getDOFsOffset());
+
+    // Copy centers
+    m_centerInFirstLink  = other.m_centerInFirstLink;
+    m_centerInSecondLink = other.m_centerInSecondLink;
+    m_centersExplicit    = other.m_centersExplicit;
+
+    this->resetAxisBuffers();
+    this->resetBuffers(other.q_previous);
+}
+
+SphericalJoint::~SphericalJoint()
+{
+}
+
+IJoint* SphericalJoint::clone() const
+{
+    return (IJoint *) new SphericalJoint(*this);
+}
+
+void SphericalJoint::setAttachedLinks(const LinkIndex _link1, const LinkIndex _link2)
+{
+    this->link1 = _link1;
+    this->link2 = _link2;
+}
+
+void SphericalJoint::setRestTransform(const Transform& _link1_X_link2)
+{
+    // Store rest rotation
+    Rotation R_rest = _link1_X_link2.getRotation();
+
+    // If centers were not explicitly provided, infer a backward-compatible default:
+    // r1 = 0, r2 = - R_rest^T p_rest so that p_rest = r1 - R_rest r2 = p_rest_input
+    if (!m_centersExplicit)
+    {
+        Position inferred_r1 = Position::Zero();
+        Position inferred_r2;
+        toEigen(inferred_r2) = - toEigen(R_rest).transpose() * toEigen(_link1_X_link2.getPosition());
+
+        m_centerInFirstLink  = inferred_r1;
+        m_centerInSecondLink = inferred_r2;
+    }
+
+    // Enforce consistency: recompute p_rest from centers and rotation
+    Position p_rest;
+    toEigen(p_rest) = toEigen(m_centerInFirstLink) - toEigen(R_rest) * toEigen(m_centerInSecondLink);
+
+    this->link1_X_link2_at_rest = Transform(R_rest, p_rest);
+
+    // Changing rest transform affects cached transforms
+    this->resetAxisBuffers();
+}
+
+const Position& SphericalJoint::centerForLink(const LinkIndex link) const
+{
+    if (link == this->link1) return m_centerInFirstLink;
+    return m_centerInSecondLink;
+}
+
+void SphericalJoint::setJointCenter(const LinkIndex link, const Position& center)
+{
+    if (link == this->link1)
+    {
+        m_centerInFirstLink = center;
+        m_centersExplicit = true;
+    }
+    else if (link == this->link2)
+    {
+        m_centerInSecondLink = center;
+        m_centersExplicit = true;
+    }
+}
```

Comment:
This does not make sense. The users just need to set one of the two, and the other will be automatically computed. See https://github.com/robotology/idyntree/blob/0991336e67400b07b1c3cf00609f7dde550b8daf/src/model/src/RevoluteJoint.cpp#L225-L239 for example. Otherwise, if the users only sets `setJointCenter(link1, center)`, the `m_centerInSecondLink` attribute will slently have a wrong value.